### PR TITLE
Contract bench

### PR DESCRIPTION
**Pull request by mike1729**
Simple contract to see gas consumption.

### CODE REVIEW

File: contracts/poseidon_host_bench/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "poseidon_host_bench"
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
```
Comment (pmikolajczyk41) [inline contracts/poseidon_host_bench/Cargo.toml:4]: :)

File: contracts/poseidon_host_bench/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+edition = "2021"
+
+[dependencies]
+ink = { version = "4.0.0", default-features = false }
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/Cargo.toml:8]:
```suggestion
ink = { version = "~4.0.0", default-features = false }
```

File: contracts/poseidon_host_bench/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
+scale-info = { version = "2.3", default-features = false, features = ["derive"], optional = true }
+baby-liminal-extension = { git = "https://github.com/Cardinal-Cryptography/aleph-node", branch = "snarkeling", default-features = false, features = ["ink-as-dependency"] }
+
+[dev-dependencies]
+ink_e2e = "4.0.0"
+
+[lib]
+path = "lib.rs"
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/Cargo.toml:19]:
```suggestion
name = "poseidon_host_bench"
path = "lib.rs"
crate-type = [
    "cdylib",
]
```

File: contracts/poseidon_host_bench/lib.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(min_specialization)]
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/lib.rs:2]: needed?

File: contracts/poseidon_host_bench/lib.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    impl PoseidonHostBench {
+        #[ink(constructor)]
+        pub fn new(max_leaves: u32) -> Self {
+            if !max_leaves.is_power_of_two() {
+                panic!("Please have 2^n leaves")
+            }
+
+            let mut bench = Self::default();
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/lib.rs:24]: bench -> `this` ?

File: contracts/poseidon_host_bench/lib.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                let right_child = self.tree_value(2 * parent + 1);
+                let parent_hash = self.hash_two(left_child, right_child);
+                self.hashes.insert(parent, &parent_hash);
+                parent /= 2;
+            }
+
+            self.next_free_leaf += 1;
+        }
+
+        fn hash_two(&self, x: MerkleHash, y: MerkleHash) -> MerkleHash {
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/lib.rs:61]: why not call it `two_to_one_hash` following the same convention

File: contracts/poseidon_host_bench/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+[features]
+default = ["std"]
+std = [
+    "ink/std",
+    "scale/std",
+    "scale-info/std",
+    "openbrush/std",
+    "baby-liminal-extension/std",
+]
+ink-as-dependency = []
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/Cargo.toml:30]: is this needed?

File: contracts/poseidon_host_bench/lib.rs (hunk 6 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/lib.rs]: done

File: contracts/poseidon_host_bench/Cargo.toml (hunk 7 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/Cargo.toml]: removed

File: contracts/poseidon_host_bench/lib.rs (hunk 5 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/lib.rs]: done

File: contracts/poseidon_host_bench/lib.rs (hunk 4 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/lib.rs]: removed

File: contracts/poseidon_host_bench/Cargo.toml (hunk 2 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/Cargo.toml]: done

File: contracts/poseidon_host_bench/Cargo.toml (hunk 1 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/Cargo.toml]: changed

File: contracts/poseidon_host_bench/.gitignore [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# Ignore build artifacts from the local tests sub-crate.
```
Comment (fbielejec) [inline contracts/poseidon_host_bench/.gitignore:1]: don't we already have one top-level gitignore to rule them all?

Comment (mike1729) [reply contracts/poseidon_host_bench/.gitignore]: It was created by `cargo contract new`. I overlooked that.

File: contracts/poseidon_host_bench/Cargo.toml (hunk 3 above)
Comment (mike1729) [reply contracts/poseidon_host_bench/Cargo.toml:27]: Added.

**Review by Yithis — APPROVED** [review]

**Review by fbielejec — APPROVED** [review]

**Pull request merged**